File: 087_ScrambleString.cpp
// Topic: Scramble String
// Description: Determine if one string is a scramble of another (binary partition + possible swap
// at each split). Approach: Recursion with memoization and early pruning using character counts.
// Worst-case exponential without memo.

#include <bits/stdc++.h>
using namespace std;

struct Key
{
    int i1, i2, len;
};

string s1, s2;
unordered_map<long long, bool> memo;

bool equalCount(int i1, int i2, int len)
{
    int cnt[26] = {};
    for (int k = 0; k < len; ++k)
    {
        cnt[s1[i1 + k] - 'a']++;
        cnt[s2[i2 + k] - 'a']--;
    }
    for (int i = 0; i < 26; ++i)
        if (cnt[i] != 0)
            return false;
    return true;
}

long long keyFor(int a, int b, int c)
{
    return ((long long)a << 40) ^ ((long long)b << 20) ^ c;
}

bool dfs(int i1, int i2, int len)
{
    long long key = keyFor(i1, i2, len);
    if (memo.count(key))
        return memo[key];
    if (s1.compare(i1, len, s2, i2, len) == 0)
        return memo[key] = true;
    if (!equalCount(i1, i2, len))
        return memo[key] = false;
    for (int cut = 1; cut < len; ++cut)
    {
        // no swap
        if (dfs(i1, i2, cut) && dfs(i1 + cut, i2 + cut, len - cut))
            return memo[key] = true;
        // swap
        if (dfs(i1, i2 + len - cut, cut) && dfs(i1 + cut, i2, len - cut))
            return memo[key] = true;
    }
    return memo[key] = false;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<pair<string, string>> tests = {{"great", "rgeat"}, {"abcde", "caebd"}, {"a", "a"}};

    for (int i = 0; i < (int)tests.size(); ++i)
    {
        s1 = tests[i].first;
        s2 = tests[i].second;
        memo.clear();
        cout << "Test " << (i + 1) << ": " << (dfs(0, 0, (int)s1.size()) ? 1 : 0) << '\n';
    }
    return 0;
}

/*
Time Complexity: High (depends on memoization) — exponential worst-case without pruning; memo
reduces repeated states. Space Complexity: O(n^3) in worst-case for memoization storage (n = string
length)
*/

File: 087_ScrambleString.py
# Topic: Scramble String
# Description: Determine if one string is a scramble of another (binary partition + possible swap at each split).
# Approach: Recursion with memoization and early pruning using character counts.

from functools import lru_cache
from collections import Counter


class Solution:
    def isScramble(self, s1: str, s2: str) -> bool:
        @lru_cache(maxsize=None)
        def dfs(i1, i2, length):
            if s1[i1:i1 + length] == s2[i2:i2 + length]:
                return True
            if Counter(s1[i1:i1 + length]) != Counter(s2[i2:i2 + length]):
                return False
            for cut in range(1, length):
                # no swap
                if dfs(i1, i2, cut) and dfs(i1 + cut, i2 + cut, length - cut):
                    return True
                # swap
                if dfs(i1, i2 + length - cut, cut) and dfs(i1 + cut, i2, length - cut):
                    return True
            return False

        return dfs(0, 0, len(s1))


if __name__ == "__main__":
    tests = [("great", "rgeat"), ("abcde", "caebd"), ("a", "a")]
    for i, (s1, s2) in enumerate(tests):
        print(f"Test {i + 1}: {Solution().isScramble(s1, s2)}")

"""
Time Complexity: Exponential worst-case without memo; memo reduces repeated states
Space Complexity: O(n^3) in worst-case for memoization storage
"""
